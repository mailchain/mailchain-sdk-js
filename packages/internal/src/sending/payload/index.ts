export * from './create';
export * from './store';
