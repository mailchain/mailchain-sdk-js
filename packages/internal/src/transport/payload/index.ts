export type { Payload } from './payload';
export { serializeAndEncryptPayload } from './payload';
export type { ContentType } from './headers';
export { SerializablePayloadHeadersImpl } from './headersSerialize';
