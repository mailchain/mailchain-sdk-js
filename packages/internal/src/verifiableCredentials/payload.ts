import { VerifiableCredential } from 'did-jwt-vc';
import { Proof } from 'did-jwt-vc/lib/types';
import { DecentralizedIdentifier } from './did';
import { TermsOfUse } from './termsOfUse';
import { CredentialSubject } from './subject';
import { W3C_CREDENTIALS_CONTEXT } from './context';

const MailchainMessagingKeyCredential = 'MailchainMessagingKeyCredential';
export const CredentialPayloadTypes = [MailchainMessagingKeyCredential] as const;
export type CredentialPayloadType = (typeof CredentialPayloadTypes)[number];

type CreateVerifiableCredentialParams = {
	type: CredentialPayloadType;
	credentialSubjects: CredentialSubject[];
	issuanceDate: Date;
	issuerId: DecentralizedIdentifier<string>;
	termsOfUse: TermsOfUse[];
	proof: Proof;
};

/**
 * Creates a verifiable credential using a provided proof instead of signature.
 */
export function createVerifiableCredential(params: CreateVerifiableCredentialParams): VerifiableCredential {
	const { type, credentialSubjects, issuanceDate, issuerId, proof, termsOfUse } = params;

	return {
		'@context': [W3C_CREDENTIALS_CONTEXT],
		credentialSubject: credentialSubjects,
		type: ['VerifiableCredential', type],
		issuanceDate: issuanceDate.toISOString(),
		issuer: {
			id: issuerId,
		},
		termsOfUse,
		proof,
	};
}
