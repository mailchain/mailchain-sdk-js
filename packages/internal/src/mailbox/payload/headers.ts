import { PublicKey } from '@mailchain/crypto';
import { createContentBuffer, MailerContent, parseMailerContentFromJSON } from '../../transport';
import {
	createOriginHeader,
	createSignatureHeader,
	headersMapFromBuffers,
	parseOriginHeader,
	parseSignatureHeader,
	SerializablePayloadHeaders,
} from '../../transport/serialization';

type ContentType = 'application/json' | 'message/x.mailchain' | 'message/x.mailchain-mailer';
/**
 * MailPayloadHeaders are the headers provide information about the payload.
 */
export type MailPayloadHeaders = {
	/**
	 * public key of sender to verify the contents
	 */
	Origin: PublicKey;
	/**
	 * used to verify un-encrypted contents
	 */
	ContentSignature: Uint8Array;

	Created: Date;

	// Standard

	/**
	 * The size of the resource, in decimal number of bytes.
	 * The Content-Length header indicates the size of the message body, in bytes, sent to the recipient.
	 */
	ContentLength: number;

	/**
	 * Indicates the media type of the resource.
	 * The ContentType representation header is used to indicate the original media type of the resource (prior to any content encoding and encryption applied for sending).
	 * In responses, a ContentType header provides the client with the actual content type of the returned content.
	 */
	ContentType: ContentType;

	/**
	 * Used to specify the compression algorithm.
	 */
	ContentEncoding: string;

	/**
	 * Used to specify the encryption algorithm.
	 */
	ContentEncryption: string;

	/**
	 * Stores original Mailer Payload when the Mail was sent with Mailer.
	 */
	MailerContent?: MailerContent;
};

const HEADER_CONTENT_ENCODING = 'Content-Encoding';
const HEADER_CONTENT_ENCRYPTION = 'Content-Encryption';
const HEADER_CONTENT_LENGTH = 'Content-Length';
const HEADER_CONTENT_SIGNATURE = 'Content-Signature';
const HEADER_CONTENT_TYPE = 'Content-Type';
const HEADER_CREATED = 'Created';
const HEADER_ORIGIN = 'Origin';
const HEADER_MAILER_CONTENT = 'Mailer-Content';

export class SerializableMailPayloadHeaders implements SerializablePayloadHeaders {
	headers: MailPayloadHeaders;

	constructor(headers: MailPayloadHeaders) {
		this.headers = headers;
	}
	static FromEncryptedMailPayloadHeaders(headers: MailPayloadHeaders): SerializableMailPayloadHeaders {
		return new this(headers);
	}

	static FromBuffer(buffer: Buffer): SerializableMailPayloadHeaders {
		const { headers } = headersMapFromBuffers(buffer, [
			HEADER_CONTENT_ENCODING,
			HEADER_CONTENT_ENCRYPTION,
			HEADER_CONTENT_LENGTH,
			HEADER_CONTENT_SIGNATURE,
			HEADER_CONTENT_TYPE,
			HEADER_CREATED,
			HEADER_ORIGIN,
		]);

		const output = {
			ContentEncoding: headers.get(HEADER_CONTENT_ENCODING)!.toString(),
			ContentEncryption: headers.get(HEADER_CONTENT_ENCRYPTION)!.toString(),
			ContentLength: Number.parseInt(headers.get(HEADER_CONTENT_LENGTH)!.toString()!),
			ContentSignature: parseSignatureHeader(headers.get(HEADER_CONTENT_SIGNATURE)!.toString()),
			ContentType: headers.get(HEADER_CONTENT_TYPE)!.toString() as ContentType,
			Created: new Date(headers.get(HEADER_CREATED)!.toString()!),
			Origin: parseOriginHeader(headers.get(HEADER_ORIGIN)!.toString()),
		} as MailPayloadHeaders;

		if (headers.get(HEADER_MAILER_CONTENT)) {
			output.MailerContent = parseMailerContentFromJSON(headers.get(HEADER_MAILER_CONTENT)!.toString());
		}

		return new this(output);
	}

	ToBuffer(): Buffer {
		const headers: string[] = [];
		headers.push(`${HEADER_CONTENT_ENCODING}: ${this.headers.ContentEncoding}`);
		headers.push(`${HEADER_CONTENT_ENCRYPTION}: ${this.headers.ContentEncryption}`);
		headers.push(`${HEADER_CONTENT_LENGTH}: ${this.headers.ContentLength}`);
		headers.push(
			`${HEADER_CONTENT_SIGNATURE}: ${createSignatureHeader(this.headers.ContentSignature, this.headers.Origin)}`,
		);
		headers.push(`${HEADER_CONTENT_TYPE}: ${this.headers.ContentType}`);
		headers.push(`${HEADER_CREATED}: ${this.headers.Created.toISOString()}`);

		headers.push(`${HEADER_ORIGIN}: ${createOriginHeader(this.headers.Origin)}`);

		if (this.headers.MailerContent) {
			headers.push(`${HEADER_MAILER_CONTENT}:  ${createContentBuffer(this.headers.MailerContent)}`);
		}

		return Buffer.from(headers.join('\r\n'), 'utf8');
	}
}
