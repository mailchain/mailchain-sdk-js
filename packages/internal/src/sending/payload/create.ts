import { KindNaClSecretKey, SignerWithPublicKey } from '@mailchain/crypto';
import { EncodingTypes } from '@mailchain/encoding';
import isArrayBuffer from 'lodash/isArrayBuffer';
import { PayloadHeaders } from '../../transport/payload/headers';
import { ContentType, Payload } from '../../transport';

export async function createPayload<C extends ContentType>(
	signerMessagingKey: SignerWithPublicKey,
	content: Buffer | Uint8Array,
	contentType: C,
): Promise<Payload<{ ContentType: C } & PayloadHeaders>> {
	return {
		Headers: {
			Origin: signerMessagingKey.publicKey,
			ContentSignature: await signerMessagingKey.sign(content),
			Created: new Date(),
			ContentLength: content.length,
			ContentType: contentType,
			ContentEncoding: EncodingTypes.Base64,
			ContentEncryption: KindNaClSecretKey,
		},
		Content: isArrayBuffer(content) ? Buffer.from(content) : content,
	};
}
