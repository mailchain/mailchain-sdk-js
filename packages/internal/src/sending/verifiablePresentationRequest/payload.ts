import { Payload } from '../../transport/payload';
import { PayloadHeaders } from '../../transport/payload/headers';

export type VerifiablePresentationHeaders = PayloadHeaders<'application/vnd.mailchain.verified-credential-request'>;

export function isVerifiablePresentationHeaders(headers: PayloadHeaders): headers is VerifiablePresentationHeaders {
	return headers.ContentType === 'application/vnd.mailchain.verified-credential-request';
}

export type VerifiablePresentationPayload = Payload<VerifiablePresentationHeaders>;

export function isVerifiablePresentationPayload(payload: Payload): payload is VerifiablePresentationPayload {
	return isVerifiablePresentationHeaders(payload.Headers);
}
