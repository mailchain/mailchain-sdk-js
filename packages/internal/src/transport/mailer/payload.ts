import { KindNaClSecretKey, SignerWithPublicKey } from '@mailchain/crypto';
import { EncodingTypes } from '@mailchain/encoding';
import { Payload } from '../payload';
import { PayloadHeaders } from '../payload/headers';
import { createContentBuffer, MailerContent } from './content';

export type MailerHeaders = PayloadHeaders<'message/x.mailchain-mailer'>;
export type MailerPayload = Payload<MailerHeaders>;

export function isMailerHeaders(headers: PayloadHeaders): headers is MailerHeaders {
	return headers.ContentType === 'message/x.mailchain-mailer' && !Object.hasOwn(headers, 'MailerContent');
}

export function isMailerPayload(payload: Payload): payload is MailerPayload {
	return isMailerHeaders(payload.Headers);
}

export type ResolvedMailerHeaders = { MailerContent: MailerContent } & MailerHeaders;
export type ResolvedMailerPayload = Payload<ResolvedMailerHeaders>;

export function isResolvedMailerHeaders(headers: PayloadHeaders): headers is ResolvedMailerHeaders {
	return headers.ContentType === 'message/x.mailchain-mailer' && Object.hasOwn(headers, 'MailerContent');
}

export function isResolvedMailerPayload(payload: Payload): payload is ResolvedMailerPayload {
	return isResolvedMailerHeaders(payload.Headers);
}

export async function createMailerPayload(
	mailerMessagingKey: SignerWithPublicKey,
	mailerContent: MailerContent,
): Promise<MailerPayload> {
	const mailerContentBuffer = Buffer.from(createContentBuffer(mailerContent));
	return {
		Headers: {
			Origin: mailerMessagingKey.publicKey,
			ContentSignature: await mailerMessagingKey.sign(mailerContentBuffer),
			Created: new Date(),
			ContentLength: mailerContentBuffer.length,
			ContentType: 'message/x.mailchain-mailer',
			ContentEncoding: EncodingTypes.Base64,
			ContentEncryption: KindNaClSecretKey,
		},
		Content: mailerContentBuffer,
	};
}
