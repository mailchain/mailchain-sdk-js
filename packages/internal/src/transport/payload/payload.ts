import { ED25519ExtendedPrivateKey } from '@mailchain/crypto';
import { CHUNK_LENGTH_1MB, encryptPayload, serialize } from '../serialization';
import { SerializablePayloadHeadersImpl } from './headersSerialize';
import { PayloadHeaders } from './headers';

/**
 * Payload.
 */
export interface Payload<H extends PayloadHeaders = PayloadHeaders> {
	/**
	 * @see defaults to {@link PayloadHeaders}
	 */
	Headers: H;
	/**
	 * Raw/data/object that are decrypted and parsed base on the the headers.
	 */
	Content: Buffer;
}

export async function serializeAndEncryptPayload(
	payload: Payload,
	payloadRootEncryptionKey: ED25519ExtendedPrivateKey,
) {
	const headers = new SerializablePayloadHeadersImpl().serialize(payload.Headers);
	return serialize(await encryptPayload(headers, payload.Content, payloadRootEncryptionKey, CHUNK_LENGTH_1MB));
}
