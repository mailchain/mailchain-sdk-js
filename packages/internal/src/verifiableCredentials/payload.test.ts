import { createVerifiableCredential } from './payload';

describe('createVerifiableCredential', () => {
	it('create - ownerOf', () => {
		const actual = createVerifiableCredential({
			credentialSubjects: [
				{
					ownerOf: {
						address: 'address',
						type: 'MailchainMessagingKey',
					},
				},
			],
			issuanceDate: new Date(2000, 1, 1),
			issuerId: 'did:mailchain:issuer',
			proof: {
				type: 'proof-type',
			},
			termsOfUse: [
				{
					id: 'terms-1',
					actions: ['action'],
					assignee: 'did:mailchain:assignee',
					assigner: 'did:mailchain:assigner',
					effect: 'Allow',
					resources: ['*'],
					type: 'HolderPolicy',
				},
			],
			type: 'MailchainMessagingKeyCredential',
		});

		expect(actual).toMatchSnapshot();
	});
});
