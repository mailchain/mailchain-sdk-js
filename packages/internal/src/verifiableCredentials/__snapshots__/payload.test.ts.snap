// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createVerifiableCredential create - ownerOf 1`] = `
{
  "@context": [
    "https://www.w3.org/2018/credentials/v1",
  ],
  "credentialSubject": [
    {
      "ownerOf": {
        "address": "address",
        "type": "MailchainMessagingKey",
      },
    },
  ],
  "issuanceDate": "2000-02-01T00:00:00.000Z",
  "issuer": {
    "id": "did:mailchain:issuer",
  },
  "proof": {
    "type": "proof-type",
  },
  "termsOfUse": [
    {
      "actions": [
        "action",
      ],
      "assignee": "did:mailchain:assignee",
      "assigner": "did:mailchain:assigner",
      "effect": "Allow",
      "id": "terms-1",
      "resources": [
        "*",
      ],
      "type": "HolderPolicy",
    },
  ],
  "type": [
    "VerifiableCredential",
    "MailchainMessagingKeyCredential",
  ],
}
`;
